document-store/src/main/java/org/hypertrace/core/documentstore/query/Query.java

```diff
@@ -34,102 +28,260 @@
  *     LIMIT 10
  * </code> can be built as <code>
  *     Query query = Query.builder()
- *         .selection(IdentifierExpression.of("col4"))
- *         .selection(
+ *         .addSelection(IdentifierExpression.of("col4"))
+ *         .addSelection(
  *             AggregateExpression.of(SUM, IdentifierExpression.of("col5")),
  *             "total")
- *         .filter(LogicalExpression.of(
+ *         .setFilter(LogicalExpression.of(
  *             RelationalExpression.of(
  *                 IdentifierExpression.of("col1"),
  *                 LT,
  *                 ConstantExpression.of(7)),
  *             AND,
  *             RelationalExpression.of(
- *             IdentifierExpression.of("col2"),
- *             NEQ,
- *             IdentifierExpression.of("col3"))))
- *         .aggregation(IdentifierExpression.of("col4"))
- *         .aggregation(IdentifierExpression.of("col6"))
- *         .aggregationFilter(
+ *                  IdentifierExpression.of("col2"),
+ *                  NEQ,
+ *                  IdentifierExpression.of("col3"))))
+ *         .addAggregation(IdentifierExpression.of("col4"))
+ *         .addAggregation(IdentifierExpression.of("col6"))
+ *         .setAggregationFilter(
  *             RelationalExpression.of(
  *                 AggregateExpression.of(SUM, IdentifierExpression.of("col5")),
  *                 GTE,
  *                 ConstantExpression.of(100)))
- *         .sortingDefinition(
+ *         .addSort(
  *             FunctionExpression.builder()
  *                 .operand(IdentifierExpression.of("col7"))
  *                 .operator(ADD)
  *                 .operand(IdentifierExpression.of("col8"))
  *                 .build(),
  *             DESC)
- *         .paginationDefinition(PaginationDefinition.of(10, 5))
+ *         .setOffset(5)
+ *         .setLimit(10)
  *         .build();
  *  </code>
  */
-@Value
 @AllArgsConstructor(access = AccessLevel.PRIVATE)
-@Builder
-public class Query {
+public final class Query {
+  private final Selection selection; // Missing selection represents fetching all the columns
+  private final Filter filter;
 
-  @NotEmpty List<@NotNull Selection> selections;
+  private final Aggregation aggregation;
+  private final Filter aggregationFilter;
```

Comment:
Just curious: Any specific reason for having `filter` and `aggregation filter` separate?

Reply:
Yup. `filter` is similar to "WHERE" clause conditions, whereas `aggregationFilter` is similar to "HAVING" clause conditions.
For the Mongo specific case, `filter` is applied before "$group" and `aggregationFilter` is applied after "$group".
Ref.: MongoQueryExecutor:67